clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -476,26 +492,48 @@ private int select(long ms) throws IOException {
     public void close(String id) {
         KafkaChannel channel = this.channels.get(id);
         if (channel != null)
-            close(channel);
+            close(channel, true);
     }
 
     /**
      * Begin closing this connection
      */
-    private void close(KafkaChannel channel) {
-        try {
-            channel.close();
-        } catch (IOException e) {
-            log.error("Exception closing connection to node {}:", channel.id(), e);
-        }
-        this.stagedReceives.remove(channel);
+    private void close(KafkaChannel channel, boolean graceful) {
+
+        channel.disconnect();
+
+        // Keep track of closed channels with pending receives so that all received records
+        // may be processed. For example, when producer with acks=0 sends some records and
+        // closes its connections, a single poll() in the broker may receive records and
+        // handle close(). When the remote end closes its connection, the channel is retained until
+        // a send fails or all outstanding receives are processed. Mute state of disconnected channels
+        // are tracked to ensure that requests are processed one-by-one by the broker to preserve ordering.
+        Deque<NetworkReceive> deque = this.stagedReceives.get(channel);
+        if (graceful && deque != null && !deque.isEmpty()) {
+            if (!channel.isMute()) {
+                addToCompletedReceives(channel, deque);
+                if (deque.isEmpty())
+                    this.stagedReceives.remove(channel);
+            }
+            closingChannels.put(channel.id(), channel);
+        } else
+            doClose(channel);
         this.channels.remove(channel.id());
         this.sensors.connectionClosed.record();
 
         if (idleExpiryManager != null)
             idleExpiryManager.remove(channel.id());
     }
 
+    private void doClose(KafkaChannel channel) {
```

Comment:
There's a slight difference in behaviour for people who close a channel directly. Before this PR, the channel would not appear in `disconnected' after a `poll` where now it will. Have you verified that this doesn't cause any issues?